src/Network/Wai/SAML2/Validation.hs

```diff
@@ -137,27 +151,40 @@ validateSAMLResponse cfg responseXmlDoc samlResponse now = do
         Left err -> throwError $ CanonicalisationFailure err
         Right result -> pure result
 
+    signature <- getSignature samlResponse
+
     -- 2. At this point we should dereference all elements identified by
     -- Reference elements inside the SignedInfo element. However, we do
     -- not currently do that and instead just assume that there is only
     -- one Reference element which targets the overall Response.
     -- We sanity check this, just in case we are wrong since we do not
     -- want an attacker to be able to exploit this.
-    let documentId = responseId samlResponse
+    documentId <- case responseSignature samlResponse of
+        Just _ -> pure $ responseId samlResponse
+        Nothing -> case responseAssertion samlResponse of
+            Just a -> pure $ assertionId a
+            Nothing -> throwError $ InvalidResponse $ userError "Assertion is missing"
     let referenceId = referenceURI
                     $ signedInfoReference
-                    $ signatureInfo
-                    $ responseSignature samlResponse
+                    $ signatureInfo signature
 
     if documentId /= referenceId
     then throwError $ UnexpectedReference referenceId
     else pure ()
 
     -- Now that we have sanity checked that we should indeed validate
-    -- the entire Response, we need to remove the Signature element
+    -- the entire Response or the Assertion, we need to remove the Signature element
     -- from it (since the Response cannot possibly have been hashed with
     -- the Signature element present). First remove the Signature element:
-    let docMinusSignature = removeSignature responseXmlDoc
+    docMinusSignature <- removeSignature <$> case responseSignature samlResponse of
+        Just _ -> pure responseXmlDoc
+        -- if a response signature is not present, assume that the assertion contains the signature
```

Comment:
When both response signature and an assertion signature are present, we could technically validate both, but I'm not sure if anyone really wants to do that